keps/140-scenario-based-simulation/README.md

```diff
@@ -43,368 +64,539 @@ Therefore, this kep proposes to define the scenario **as CRD**. All clients, inc
-- change the scheduler configuration via simulator API.
-- create/delete/edit any resources.
-
-And all resources created before starting a scenario are deleted at the start of the scenario,
-so that they don't affect the simulation results.
-
-### User Stories 
+The scenario controller only updates status.scenarioResult in Scenario resource when proceeding to the next ScenarioStep.
+This is because kube-apiserver will be so busy if the controller update status.scenarioResult everytime it updated,
+especially when the size of Scenario is so big.
 
-#### Story 1
+> etcd is designed to handle small key value pairs typical for metadata. Larger requests will work, but may increase the latency of other requests. By default, the maximum size of any request is 1.5 MiB. This limit is configurable through --max-request-bytes flag for etcd server.
+https://etcd.io/docs/v3.4/dev-guide/limit/#request-size-limit
 
-The company has added many features into scheduler via some custom plugins, 
-and they want to make sure that their expansions are working as expected and has not negatively impacted the scheduling results.
+For example, when using Scenario for accurate benchmark testing, users may want to reduce the request to update Scenario for kube-apiserver as much as possible.
 
-##### Solution
+We can add a new configuration environment variable `UPDATE_SCENARIO_RESULTS_STRATEGY` and define some strategies like:
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=AtMovingNextStep`: It's default value. update status.scenarioResult in Scenario resource when proceeding to the next ScenarioStep.
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=OnPause`: update status.scenarioResult in Scenario resource when the Scenario's phase becomes `Paused`, `Succeeded` or `Failed`.
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=OnDone`: update status.scenarioResult in Scenario resource when the Scenario's phase becomes `Succeeded` or `Failed`.
 
-They can define appropriate scenario and analize the results.
+We can discuss it after all the implementation of Scenario is done.
 
-#### Story 2
+#### The case kube-apiserver has Scenarios when the scenario controller starts to run
 
-The users want to see how their customized scheduler behaves in the worst case scenario.
+When the scenario controller is started and finds the Scenario which phase is "Running", the controller changes the status "Failed" with updating the `.status.message` like "the controller restarted while the Scenario was running".
 
-##### Solution
+#### Prohibitions and restrictions
 
-Even when a scenario is running, users can add events to that scenario. 
-So, in this case, they can add events that are most worst case for the scheduler by looking at the simulation results and the resources status at that point.
+The scenarios run one by one, and multiple scenarios are never run simultaneously. 
+That means the scenario controller will run the following Scenario after the current running Scenario becomes "Failed" or "Succeeded".
```

Comment:
If a user wanted to run hundreds of Scenarios, how to order them??  Is there any interface to specify the order for executing Scenarios??

Reply:
> If a user wanted to run hundreds of Scenarios 

I'm thinking they are gonna run in the order in which they are created.

> Is there any interface to specify the order for executing Scenarios??

I'd say we don't need to consider it at least for now, but it may be useful to have.